nixos/mautrix-telegram: switch to using static user for automated registration of appservice

thefeiter opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done
1. Create a static user
2. Add config option registerToSynapse and permission management to read registration file
3. enable env replacement in config json and automatically add generated AS and HS tokens to config

This helps solving the Issue described [here](https://discourse.nixos.org/t/setting-up-the-as-token-for-mautrix-telegram-matrix-bridge/34025/2)

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] n.a. ~~Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)~~
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md). afaik :)

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

I have tested the changes in a build-vm environment and the generation and registration of the appserveice work fine!

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

thefeiter commented:
If backporting means, making the change available in an older release, this would be nice for at least 24.11 I think.

thefeiter commented:
Infos for release notes are too insignificant I guess.

thefeiter commented:
I don't know of any tests belonging to this module.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/setting-up-the-as-token-for-mautrix-telegram-matrix-bridge/34025/6

matthiasbeyer reviewed (approved):
I see nothing that would concern me. I would wait one day for either @Stunkymonkey or @pacien to say something and, if they do not respond, merge.

matthiasbeyer commented:
@thefeiter can you please rebase this and remove the merge commit?

thefeiter committed:
nixos/mautrix-telegram: switch to using static user for automated registration of appservice
Like for matrix-whatsapp use a static user so that the registration file can be automatically shared with synapse.
This also includes the registerToSynapse config option.

thefeiter commented:
oof, me and using git...

matthiasbeyer commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 389679`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>

matthiasbeyer commented:
As neither @Stunkymonkey nor @pacien replied, I am gonna merge this now.

Stunkymonkey commented:
> As neither @Stunkymonkey nor @pacien replied, I am gonna merge this now.

@matthiasbeyer  sorry. i am a bit short on time. But it looks good to me. thanks for the contribution.

on little thing: using `umask` is not encouraged #121293

matthiasbeyer commented:
> using `umask` is not encouraged #121293

@thefeiter care to fix this up in a follow-up PR? Just for your training, you know? :laughing:

krishukr commented:
Hi, just updated my NixOS from 24.11 to 25.05. I found that this PR introduced a breaking change for mautrix-telegram service.

Following the official mautrix-bridges [documentation](https://docs.mau.fi/bridges/python/setup.html?bridge=telegram#production-setup) and NixOS Wiki [page](https://wiki.nixos.org/wiki/Matrix#mautrix-telegram), one would already set up a `telegram-registration.yaml` somewhere possibly *not* in `/var/lib/mautrix-telegram` per guides in original [docs](https://docs.mau.fi/bridges/general/registering-appservices.html).  However, the auto-generated and installed application service registration file, located in `/var/lib/mautrix-telegram`, would cause `ConfigError` due to ID reuse. 

```
synapse.config.appservice: [main] Cannot reuse ID across application services: telegram (files: /var/lib/matrix-synapse/telegram-registration.yaml, /var/lib/mautrix-telegram/telegram-registration.yaml)
```

Automated configuration for mautrix-telegram would be beneficial. But a release note addressing this issue or an automated check to prevent it is also necessary.
